api/envoy/extensions/filters/common/set_filter_state/v3/value.proto

```diff
@@ -0,0 +1,60 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.common.set_filter_state.v3;
+
+import "envoy/config/core/v3/substitution_format_string.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.common.set_filter_state.v3";
+option java_outer_classname = "ValueProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/common/set_filter_state/v3;set_filter_statev3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Set-Filter-State filter state value]
+
+// A filter state key and value pair.
+// [#next-free-field: 6]
+message FilterStateValue {
+  enum SharedWithUpstream {
+    // Object is not shared with the upstream internal connections.
+    NONE = 0;
+
+    // Object is shared with the upstream internal connection.
+    ONCE = 1;
+
+    // Object is shared with the upstream internal connection and any internal connection upstream from it.
+    TRANSITIVE = 2;
+  }
+
+  oneof key {
+    option (validate.required) = true;
+
+    // Filter state object key. The key is expected to be registered via an object factory, see
+    // :ref:`the well-known filter state keys <well_known_filter_state>`.
+    string object_key = 1 [(validate.rules).string = {min_len: 1}];
```

Comment:
Maybe we can add new `factory_key` in the future to search the factory. Then the users could set multiple same type objects to the filter state with different object key.

We do lots of similar thing in other positions. `index (key)` is not bound to `type (factory)`.

Reply:
Yes, we can add indirection. @ohadvano also asked to add a generic `int_key` or `string_key` for standard types. I made it oneof to support this extensibility.